tokenFactory.go

```diff
@@ -0,0 +1,117 @@
+ * limitations under the License.
+ *
+ */
+
+package main
+
+import (
+	"context"
+	"errors"
+	"net/http"
+
+	"emperror.dev/emperror"
+	"github.com/dgrijalva/jwt-go"
+	"github.com/xmidt-org/bascule"
+	"github.com/xmidt-org/bascule/basculehttp"
+	"github.com/xmidt-org/bascule/key"
+)
+
+const (
+	jwtPrincipalKey = "sub"
+)
+
+// AccessLevelBearerTokenFactory extends basculehttp.BearerTokenFactory by injecting an access level attribute to the jwt token. How the level is generated is
+// left to the user.
+type AccessLevelBearerTokenFactory struct {
+	DefaultKeyID      string
+	Resolver          key.Resolver
+	Parser            bascule.JWTParser
+	Leeway            bascule.Leeway
+	AccessLevelConfig AccessLevelConfig
+}
+
+func (a AccessLevelBearerTokenFactory) GetDefaultKeyID() string {
+	if a.DefaultKeyID != "" {
+		return a.DefaultKeyID
+	}
+	return "current"
+}
+
+// ParseAndValidate expects the given value to be a JWT with a kid header.  The
+// kid should be resolvable by the Resolver and the JWT should be Parseable and
+// pass any basic validation checks done by the Parser.  If everything goes
+// well, a Token of type "jwt" is returned.
+func (a AccessLevelBearerTokenFactory) ParseAndValidate(ctx context.Context, _ *http.Request, _ bascule.Authorization, value string) (bascule.Token, error) {
+	if len(value) == 0 {
+		return nil, errors.New("empty value")
+	}
+
+	leewayclaims := bascule.ClaimsWithLeeway{
+		MapClaims: make(jwt.MapClaims),
+		Leeway:    a.Leeway,
+	}
+
+	jwsToken, err := a.Parser.ParseJWT(value, &leewayclaims, defaultKeyfunc(ctx, a.GetDefaultKeyID(), a.Resolver))
+	if err != nil {
+		return nil, emperror.Wrap(err, "failed to parse JWS")
+	}
+	if !jwsToken.Valid {
+		return nil, basculehttp.ErrorInvalidToken
+	}
+
+	claims, ok := jwsToken.Claims.(*bascule.ClaimsWithLeeway)
+
+	if !ok {
+		return nil, emperror.Wrap(basculehttp.ErrorUnexpectedClaims, "failed to parse JWS")
+	}
+
+	claimsMap, err := claims.GetMap()
+	if err != nil {
+		return nil, emperror.WrapWith(err, "failed to get map of claims", "claims struct", claims)
+	}
+
+	jwtClaims := bascule.NewAttributes(claimsMap)
+
+	accessLevelResolver := a.AccessLevelConfig.GetResolver()
+	claimsMap[a.AccessLevelConfig.GetAttributeKey()] = accessLevelResolver(jwtClaims)
+
+	principalVal, ok := jwtClaims.Get(jwtPrincipalKey)
+	if !ok {
+		return nil, emperror.WrapWith(basculehttp.ErrorInvalidPrincipal, "principal value not found", "principal key", jwtPrincipalKey, "jwtClaims", claimsMap)
+	}
+	principal, ok := principalVal.(string)
+	if !ok {
+		return nil, emperror.WrapWith(basculehttp.ErrorInvalidPrincipal, "principal value not a string", "principal", principalVal)
+	}
+
+	return bascule.NewToken("jwt", principal, jwtClaims), nil
+}
+
+// TODO: maybe we should have bascule export something like this
+var defaultKeyfunc = func(ctx context.Context, defaultKeyID string, keyResolver key.Resolver) jwt.Keyfunc {
```

Comment:
I was thinking to avoid code duplication. It seems like a pretty stable component that can be used by different implementations of token factories.

Reply:
Wouldn't this be accomplished with a function declaration instead of a variable? The only reason I can see to make this a variable is to replace the functionality in unit tests, which doesn't seem needed.